### PR TITLE
Thread Sanitization

### PR DESCRIPTION
**Pull request by Ashod**
Some thread-safety fixes with the help of the thread sanitizer.

- wsd: signal handler cannot log
- wsd: test: better test timeout thread management
- wsd: thread-safe socket shutdown flag
- wsd: thread-safe kit pid
- wsd: signal handlers must preserve errno

### CODE REVIEW

File: common/Unit.cpp
Comment (Ashod) [inline common/Unit.cpp:538]: Manually unlocking here was wrong because 1) we didn't take the lock and 2) the Mutex might already be unlocked.

Both cases are invalid.

**Review by mmeeks — COMMENTED** [review]
All looks fine to me - another review needed before merge to master =)

File: common/SigUtil.cpp
Comment (vmiklos) [inline common/SigUtil.cpp:298]: Nice, joke, but perhaps go with something like oldErrno or origErrno to be more readable?

**Review by vmiklos — APPROVED** [review]
Looks OK to me, please wait for CI and then I'm fine to merge this. See inline comment for a nit. Thanks.

**Pull request merged**
